src/types.rs

```diff
@@ -956,12 +957,79 @@ impl Types {
                 ref body,
                 var_kind,
             } => {
-                use VarKind::*;
+                let mut constr: HashSet<Ident> = HashSet::new();
+                // TODO: optimization: might be faster if maybe_constr is a Vec, and since it only
+                // ever gets merged into `constr`, this won't affect semantics.
+                let mut maybe_constr: HashSet<Ident> = HashSet::new();
+                let mut to_be_checked: Vec<&Types> = vec![body];
+                while let Some(tys) = to_be_checked.pop() {
+                    match &tys.types {
```

Comment:
This is a good thought. Hmm... I'm not sure if you meant something else by visitor pattern, but I tried implementing it as an iterator, and in order to allow the shadowed variable check, it had to look something like this:

```rust
pub struct TypesIterator<'a> {
    pub(crate) todo: Vec<&'a Types>,
    pub(crate) should_recurse: Option<Box<dyn Fn(&'a Types) -> bool>>,
}

impl<'a> Iterator for TypesIterator<'a> {
    type Item = &'a Types;

    fn next(&mut self) -> Option<Self::Item> {
        self.todo.pop().map(|tys| {
            let should_recurse = match &self.should_recurse {
                None => true,
                Some(f) => f(tys),
            };
            if should_recurse {
                match &tys.types {
                    TypeF::Arrow(s, t) => {
                        self.todo.push(s);
                        self.todo.push(t);
                    }
                    TypeF::Forall { body, .. } => {
                        self.todo.push(body);
                    }
                    TypeF::Record(rrows) => {
                        for ritem in rrows.iter() {
                            if let RecordRowsIteratorItem::Row(row) = ritem {
                                self.todo.push(row.types);
                            }
                        }
                    }
                    TypeF::Dict {
                        type_fields,
                        flavour,
                    } => {
                        // XXX: is this the right semantics? I'm not sure whether we should recurse
                        // into contracts or not
                        if flavour == &DictTypeFlavour::Type {
                            self.todo.push(type_fields);
                        }
                    }
                    TypeF::Array(t) => {
                        self.todo.push(t);
                    }
                    _ => (),
                }
            }
            tys
        })
    }
}

impl Types {
    pub fn iter(&self, should_recurse: Option<impl Fn(&Types) -> bool + 'static>) -> TypesIterator {
        TypesIterator {
            todo: vec![self],
            should_recurse: should_recurse.map(|x| Box::new(x) as Box<dyn Fn(&Types) -> bool>),
        }
    }
}
```

And then you call it like
```rust
                let var_ = var.clone();
                let ignore_shadowed = move |tys: &Types| match &tys.types {
                    TypeF::Forall { var, .. } if var == &var_ => false,
                    _ => true,
                };
                for tys in body.iter(Some(ignore_shadowed)) {
                    ...
                }
```

Advantages:
- clear separation of concerns

Problems:
- It's not clear that this is actually general enough. What if someone wants to only recurse into the domain of a function type for some reason?
- Dealing with these closure types seems really thorny. It took me a while to get the types right, and even then I'm not sure how to get it to work if you pass in `None`, because it needs a concrete type there.

One alternative is a much more specialized function, like `iter_forall(body: &Types, var: Ident)` which takes into account shadowing. But that feels really specific...

Reply:
Yup, I was thinking of something like this. Even if it's not reused, I feel it separates concerns and makes the code more readable (the call site to this method is only concerned with the logic of the processing; if you want to see how the iteration work, you can then go to the definition of `iter()`).

> It's not clear that this is actually general enough. What if someone wants to only recurse into the domain of a function type for some reason?

This is a good question, which is hard to answer now. But, what we can do is to give a specific name to this function, make the documentation clear enough. If one day we need something more general, it shouldn't be too hard to do, and to replace the function you wrote by a specialized call to the general one.

> Dealing with these closure types seems really thorny. It took me a while to get the types right, and even then I'm not sure how to get it to work if you pass in None, because it needs a concrete type there.

I'm not sure to see why we need `'static` here, by the way. And we can probably do with `FnMut` instead of `Fn`. But, as far as the `Option` is concerned, I would just get rid of it. `None` is semantically equivalent to `|_| true`, which is really not that longer to write for the caller, while as you mentioned, it is annoying with generics (and require an additional pattern matching). I'm relatively confident that the optimized code will be as efficient with `|_| true` as with `None`, thanks to monorphization.